src/main/java/bio/terra/service/snapshot/flight/duos/SnapshotUpdateDuosDatasetFlight.java

```diff
@@ -0,0 +1,58 @@
+package bio.terra.service.snapshot.flight.duos;
+
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.model.DuosFirecloudGroupModel;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.duos.DuosDao;
+import bio.terra.service.duos.DuosService;
+import bio.terra.service.job.JobMapKeys;
+import bio.terra.service.snapshot.SnapshotDao;
+import bio.terra.stairway.Flight;
+import bio.terra.stairway.FlightMap;
+import java.util.UUID;
+import org.springframework.context.ApplicationContext;
+
+/**
+ * This flight links a snapshot to a DUOS dataset for the purposes of syncing its authorized users
+ * as snapshot readers via a TDR-managed Firecloud group. It supports linking, unlinking, and
+ * updating a snapshot's existing link (unlinking + linking).
+ */
+public class SnapshotUpdateDuosDatasetFlight extends Flight {
+
+  public SnapshotUpdateDuosDatasetFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    ApplicationContext appContext = (ApplicationContext) applicationContext;
+    DuosDao duosDao = appContext.getBean(DuosDao.class);
+    DuosService duosService = appContext.getBean(DuosService.class);
+    IamService iamService = appContext.getBean(IamService.class);
+    SnapshotDao snapshotDao = appContext.getBean(SnapshotDao.class);
+
+    AuthenticatedUserRequest userReq =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    UUID snapshotId = inputParameters.get(JobMapKeys.SNAPSHOT_ID.getKeyName(), UUID.class);
+    String duosId = inputParameters.get(SnapshotDuosMapKeys.DUOS_ID, String.class);
+    DuosFirecloudGroupModel firecloudGroupPrev =
+        inputParameters.get(
+            SnapshotDuosMapKeys.FIRECLOUD_GROUP_PREV, DuosFirecloudGroupModel.class);
+
+    boolean linking = (duosId != null);
+    boolean unlinking = (firecloudGroupPrev != null);
```

Comment:
I think it's possible for both duoId & firecloudGroupPrev to be null when a user hits the unlinkDuosDatasetFromSnapshot endpoint for a snapshot that doesn't already have a linked duosID. 
So, I'm thinking we might want to:
(a) Make sure the UpdateSnapshotDuosFirecloudGroupIdStep can still handle this case since it would still be hit
(b) Provide feedback to user that the snapshot was not linked to anything

Reply:
Yes, you're right that this is possible!  I think there's no harm in allowing this case, and the response will indicate that nothing was unlinked as well as nothing linked.  I will update the unit tests to specifically validate this case.

Reply:
Ah, I was thinking that getDuosFirecloudGroupId would fail, but looks like you were already handling the case whether the group is null! Nice!

Reply:
I added an additional test -- `UpdateSnapshotDuosFirecloudGroupIdStepTest.testDoAndUndoUnlinkNoOp` :)